### ISSUE TITLE
Custom scope modules are added too late

### ISSUE THREAD

**Issue by Tim203** [avaje/avaje-inject#295]
This results in the build failing to compile since the custom module class can't be found.
Note that this only seems to fail when a different package is used than the main module (the main module always shows up.)
[Link to the repo](https://github.com/Tim203/java-tests)

Error message:
```
C:\...\src\main\java\net\tim203\tests\Main.java:4: error: cannot find symbol
import net.tim203.tests.hi.CustomModule;
                          ^
  symbol:   class CustomModule
  location: package net.tim203.tests.hi
```

**Comment by SentryMan** [comment]
hmm, it only seems to happen with gradle

**Comment by SentryMan** [comment]
Yeah, don't even know where to start with this. @rbygrave do you have any ideas?

**Comment by rbygrave** [comment]
It works as expected with maven (and IntelliJ with Maven only). Note to test that with maven we do need to temporarily move all the gradle files out of the project.

**Comment by SentryMan** [comment]
when I converted his project to maven I got the same error

**Comment by rbygrave** [comment]
> It works as expected with maven

Ah, actually it doesn't work with maven.

> ideas

My theory is that this issue is because the module is written in the last round of processing. That is, because the module is written in the last round it then does not exist as an element that can be used in in other `main` source (but it can be used in `test` source).

Now, we build the module in the last round as it has the ordering logic in it. It my theory is correct we need to split this into 2 classes - one with the ordering that MUST be written in the last round and another which should be written early (which we can then use in `source main`).

**Comment by rbygrave** [comment]
@mikehearn When using custom module scopes in `src/main` do you create those custom module instance via reflection?  Or are they in a seperate maven/gradle module perhaps? I'm wondering if you hit this issue?

**Comment by rbygrave** [comment]
Refer to: https://bugs.openjdk.org/browse/JDK-8256826

**Comment by rbygrave** [comment]
Ok, a couple of paths forward:

Option 1) Write the custom modules earlier / not in the final processing round.  This should be all fine as long as we don't have other annotation processors also generating classes that use the custom scope.

Note that we don't want to do this for the "default scope" which uses `@jakarta.inject.Singleton` as we have for example avaje-http which generates controller adapters for the "default scope". The thought here is that it is probably reasonable to add this limitation on custom scopes though - the limitation being that we can't have other annotation processors also generating classes for the custom scope. This seems like a reasonable limitation to add/impose.


Option 2) In theory this could be a JDK bug that is actually fixed in Java 11. The bug referenced above does reproduce for Java 17 but for Java 19 I can get it to work as long as we create the JavaFileObject before the last processing round. However, I was unable to confirm the fix via maven-compiler-plugin 3.11.0 and this test project.  More digging is required to see the difference between raw javac and the maven-compiler-plugin.


My gut says we go with Option 1.

**Comment by rbygrave** [comment]
We have released 9.0-RC2 which includes the fix for this issue.
